node.go

```diff
@@ -86,21 +87,32 @@ func PrefixForCidVersion(version int) (cid.Prefix, error) {
 
 // CidBuilder returns the CID Builder for this ProtoNode, it is never nil
 func (n *ProtoNode) CidBuilder() cid.Builder {
-	if n.builder == nil {
-		n.builder = v0CidPrefix
+	n.mu.RLock()
+	cb := n.cidBuilder
+	n.mu.RUnlock()
+
+	if cb != nil {
+		return cb
 	}
-	return n.builder
+
+	n.mu.Lock()
```

Comment:
are we OK with this lock being separate from the original lock? e.g. two goroutines might get to setting cidBuilder to v0CidPrefix. I guess we're OK with that as both would write the same value.

(speaking of which, why is it written to a field at all, then?)

Reply:
actually, this might be a "logic race" with SetCidBuilder; if I call both, and SetCidBuilder wins the race to the mutex first, then CidBuilder here can later override the custom builder with v0CidPrefix.

Reply:
>  why is it written to a field at all, then

I just copied what we did before, it does make little sense... Alternatively: yes, I could just make "private race-prone" versions and lock the entire public method end-to-end, to avoid stuff like above. I'll probably just do that in the next hour or so.